lldb/tools/lldb-dap/DAP.cpp

```diff
@@ -1310,103 +1396,161 @@ void DAP::EventThread() {
                   "module", ModuleEventBody{std::move(p_module).value(),
                                             ModuleEventBody::eReasonChanged}});
             } else if (!remove_module) {
-              modules.insert(module_id);
-              Send(protocol::Event{
+              dap_instance->modules.insert(module_id);
+              dap_instance->Send(protocol::Event{
                   "module", ModuleEventBody{std::move(p_module).value(),
                                             ModuleEventBody::eReasonNew}});
             }
           }
+        } else if (event_mask & lldb::SBTarget::eBroadcastBitNewTargetSpawned) {
+          auto target = lldb::SBTarget::GetTargetFromEvent(event);
+          auto target_index = debugger.GetIndexOfTarget(target);
+
+          // Set the shared debugger for GPU processes
+          DAPSessionManager::GetInstance().SetSharedDebugger(target_index,
+                                                             debugger);
+
+          // We create "attachCommands" that will select the target that already
+          // exists in LLDB. The DAP instance will attach to this already
+          // existing target and the debug session will be ready to go.
+          llvm::json::Object attach_config;
+          llvm::json::Array attach_commands;
+
+          attach_commands.push_back(llvm::formatv("target list").str());
+          attach_commands.push_back(
+              llvm::formatv("target select {0}", target_index).str());
+
+          // If we have a process name, add command to attach to the same
+          // process name
+
+          attach_config.try_emplace("type", "lldb");
+          attach_config.try_emplace("name", "GPU Session");
+          attach_config.try_emplace("attachCommands",
+                                    std::move(attach_commands));
+          attach_config.try_emplace("targetIdx", target_index);
+
+          // 2. Construct the main 'startDebugging' request arguments.
+          llvm::json::Object start_debugging_args;
+          start_debugging_args.try_emplace("request", "attach");
+          start_debugging_args.try_emplace("configuration",
+                                           std::move(attach_config));
+
+          // Send the request. Note that this is a reverse request, so you don't
+          // expect a direct response in the same way as a client request.
+          SendReverseRequest<LogFailureResponseHandler>(
+              "startDebugging", std::move(start_debugging_args));
         }
       } else if (lldb::SBBreakpoint::EventIsBreakpointEvent(event)) {
+        lldb::SBBreakpoint bp =
+            lldb::SBBreakpoint::GetBreakpointFromEvent(event);
+        if (!bp.IsValid())
+          continue;
+
+        lldb::SBTarget event_target = bp.GetTarget();
+
+        // Find the DAP instance that owns this target
+        DAP *dap_instance =
+            DAPSessionManager::GetInstance().FindDAPForTarget(event_target);
```

Comment:
FindDAPForTarget is used very often, you might as well create a static override that will under the hood invoke GetInstance(), just to simplify the callers